[4.1] PHP 8.1 deprecation notices libraries/src

frostmakk opened the pull request:
Pull Request for Issue # .

### Summary of Changes
Patching PHP 8.1 deprecation notices where null is provided instead of 'something'. 


### Testing Instructions
Code review or checking logs.


### Actual result BEFORE applying this Pull Request
Deprecation notices


### Expected result AFTER applying this Pull Request
No deprecation notices from these files.


### Documentation Changes Required
Probably not.

frostmakk committed:
PHP 8.1 deprecation notices libraries/src

frostmakk committed:
Update ComponentlayoutField.php

frostmakk committed:
coding standard

ghazal commented:
I have tested this item :white_check_mark: successfully on ef22f70e3fe4fa0a683b077230b20a0b77d21a62<br /><br />I chose "Tested successfully" because it does the job, ie showing Deprecated Notices distributed all over admin.

Here is what I collected:

Deprecated: strtolower(): Passing null to parameter #1 ($string) of type string is deprecated in /administrator/components/com_config/src/Controller/DisplayController.php on line 53
  
Deprecated: Automatic conversion of false to array is deprecated in /libraries/src/Form/Field/RulesField.php on line 181
      
Deprecated: Function strftime() is deprecated in /libraries/src/Form/Field/CalendarField.php on line 277

Deprecated: htmlspecialchars(): Passing null to parameter #1 ($string) of type string is deprecated in /libraries/src/MVC/View/HtmlView.php on line 254

Deprecated: htmlspecialchars(): Passing null to parameter #1 ($string) of type string is deprecated in /libraries/src/Layout/BaseLayout.php on line 119

Deprecated: substr(): Passing null to parameter #1 ($string) of type string is deprecated in /administrator/components/com_fields/src/Model/GroupModel.php on line 323<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36355">issues.joomla.org/tracker/joomla-cms/36355</a>.</sub>

frostmakk committed:
Adding ../Table/Module.php

frostmakk committed:
Adding 6 more.

alikon commented:
I have tested this item :white_check_mark: successfully on b5d53befdbadd23abbb5348d264f7e67061b1681<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36355">issues.joomla.org/tracker/joomla-cms/36355</a>.</sub>

alikon commented:
@ghazal can yuo please re-test this 

and if you can

- #36362
- #36360

thx

frostmakk committed:
Update PunycodeHelper.php

frostmakk committed:
Empty line

HLeithner commented on libraries/src/Application/AdministratorApplication.php:
```diff
@@ -386,7 +386,11 @@ public function login($credentials, $options = array())
 			$lang = $this->input->getCmd('lang');
-			$lang = preg_replace('/[^A-Z-]/i', '', $lang);
+
+			if (isset($lang))
```
`isset` is would be always true, add a second parameter to the `getCmd` call

HLeithner commented on libraries/src/Application/AdministratorApplication.php:
```diff
@@ -386,7 +386,11 @@ public function login($credentials, $options = array())
+
+			if (isset($lang))
+			{
+				$lang = preg_replace('/[^A-Z-]/i', '', $lang);
```
What are you trying to do here? If I'm not wrong then `getCmd` above does exactly the same.

frostmakk committed:
Merge branch 'joomla:4.0-dev' into php81-lib

HLeithner commented on libraries/src/Application/CMSApplication.php:
```diff
@@ -199,6 +199,11 @@ public function checkSession()
 		// Don't add empty messages.
+		if ($msg === null)
+		{
+			$msg = '';
```
makes no sense. 2 solutions

1. make an early return
2. combine the 2 if blocks to `if ($msg === null || trim($msg) === '')`

HLeithner commented on libraries/src/Date/Date.php:
```diff
@@ -119,6 +119,12 @@ public function __construct($date = 'now', $tz = null)
+		// If called with null $date
+		if ($date === null)
+		{
+			$date = date('now');
```
the parent class expects a string as first parameter
```suggestion
			$date = 'now';
```

HLeithner commented on libraries/src/Filesystem/Path.php:
```diff
@@ -216,7 +216,10 @@ public static function clean($path, $ds = DIRECTORY_SEPARATOR)
 		}
 
-		$path = trim($path);
+		if (isset($path))
+		{
+			$path = trim($path);
+		}
 
 		if (empty($path))
```
```suggestion
		if (empty($path) || trim($path) === '')
```

HLeithner commented on libraries/src/Filter/OutputFilter.php:
```diff
@@ -47,7 +47,7 @@ public static function linkXHTMLSafe($input)
 	public static function stringJSSafe($string)
 	{
-		$chars = preg_split('//u', $string, null, PREG_SPLIT_NO_EMPTY);
+		$chars = preg_split('//u', $string, 0, PREG_SPLIT_NO_EMPTY);
```
```suggestion
		$chars = preg_split('//u', $string, -1, PREG_SPLIT_NO_EMPTY);
```
Use the same default value as defined in the documentation.

HLeithner commented on libraries/src/Filter/OutputFilter.php:
```diff
@@ -81,6 +81,11 @@ public static function stringJSSafe($string)
 	public static function stringURLSafe($string, $language = '')
 	{
 		// Remove any '-' from the string since they will be used as concatenaters
```
Comment is on the wrong line

HLeithner commented on libraries/src/Form/Field/ComponentlayoutField.php:
```diff
@@ -75,7 +75,11 @@ protected function getInput()
 			$template_style_id = $this->form->getValue('template_style_id');
-			$template_style_id = (int) preg_replace('#\W#', '', $template_style_id);
+
+			if (isset($template_style_id))
```
if we want to check for `null` we always use `$x === null`

HLeithner commented on libraries/src/Form/Field/ModulelayoutField.php:
```diff
@@ -76,7 +76,11 @@ protected function getInput()
 			$template_style_id = $this->form->getValue('template_style_id');
-			$template_style_id = (int) preg_replace('#\W#', '', $template_style_id);
+
+			if (isset($template_style_id))
```
if we want to check for `null` we always use `$x === null`

HLeithner commented on libraries/src/MVC/Controller/BaseController.php:
```diff
@@ -315,7 +315,7 @@ public static function getInstance($prefix, $config = array())
 
 		// Get the controller class name.
-		$class = ucfirst($prefix) . 'Controller' . ucfirst($type);
+		$class = ucfirst($prefix) . 'Controller' . ucfirst($type ?? '');
```
`$type` is set in the code 5 line above to null, change this line...

HLeithner commented on libraries/src/MVC/Controller/BaseController.php:
```diff
@@ -709,7 +709,10 @@ public function execute($task)
 		$this->task = $task;
 
-		$task = strtolower($task);
+		if (isset($task))
```
if we want to check for `null` we always use `$x === null`

frostmakk replied:
I thought we were checking for anything but null here ???

HLeithner replied:
then use `!==`

HLeithner commented on libraries/src/String/PunycodeHelper.php:
```diff
@@ -132,6 +132,8 @@ public static function urlToUTF8($uri)
 	{
 		if (empty($uri))
 		{
+			$uri = '';
+
 			return;
```
```suggestion
			return '';
```
no comment...

frostmakk committed:
Update libraries/src/String/PunycodeHelper.php

Co-authored-by: Harald Leithner <leithner@itronic.at>

frostmakk committed:
Update libraries/src/Filter/OutputFilter.php

Co-authored-by: Harald Leithner <leithner@itronic.at>

frostmakk committed:
Update libraries/src/Filesystem/Path.php

Co-authored-by: Harald Leithner <leithner@itronic.at>

ghazal commented:
@alikon 
I'll test this again as soon as @HLeithner requested changes are solved.
I have begun to test #36362 and #36360 via patchtester, not finished yet.
Right now, I got mainly deprecation notices around strftime(), but I guess this is going to be corrected.

frostmakk committed:
Update libraries/src/Date/Date.php

Co-authored-by: Harald Leithner <leithner@itronic.at>

frostmakk committed:
Requested changes

frostmakk committed:
Merge branch 'php81-lib' of https://github.com/frostmakk/joomla-cms into php81-lib

frostmakk committed:
Moved comment

810 commented:
I have tested this item :white_check_mark: successfully on 15a66838f112b92f751707d429bd5796b9e6bda5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36355">issues.joomla.org/tracker/joomla-cms/36355</a>.</sub>

PhilETaylor commented:
@richard67 this PR has 2 tests - can it please be RTC now so that it can be merged quicker, to stop repeated issues being opened about it

richard67 commented:
@PhilETaylor After @alikon 's there have been made changes which were not just code style, that's why only one test is counted right now if you check in the issue tracker or here at the bottom in the CI checks.

@alikon Could you test again? Thanks in advance. I am at work now so I can't test anything myself.

PhilETaylor commented:
ah ok missed that - some controversial changes here, but nothing that's going to break Joomla.

ghazal commented:
I have tested this item :white_check_mark: successfully on 15a66838f112b92f751707d429bd5796b9e6bda5<br /><br />From a fresh install of 410 beta3.
BTW, I can apply this PR via patchtester only if I don't apply this one:

[4.0] PHP 8 Ensure that we have a valid date #36396 

@PhilETaylor 
Thank you for your list of PRs.

[4][com_finder] - php 8.1 serializable-deprecated with unit test #36482
[4.0 - PHP8.1] Remove call to deprecated strftime - with unit test #36469
[4.0] Fix PHP 8.1 deprecation - strtolower null param #36397
[4.0] PHP 8 Ensure that we have a valid date #36396
[4.0] PHP 8.1 deprecation notices libraries/src #36355
PHP 8.1 deprecation fixes #36306


<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36355">issues.joomla.org/tracker/joomla-cms/36355</a>.</sub>

richard67 commented:
> From a fresh install of 410 beta3.

@ghazal Hmm, this PR is for the 4.0-dev branch and so not for 4.1. For lucky circumstance (no other changes on these files meanwhile between 4.0-dev and 4.1-dev) it might be that the PR worked for you. But in general it's not a good idea to test a pull request on a version based on a different branch (4.1-dev) than the target branch of that pull request (4.0-dev).

I am not sure now if I shall count your test and so set the PR to RTC.

Maybe you can test again on a 4.0.x (latest 4.0 nightly or latest 4.0.5 release? That would be great.

Or maybe @alikon can test again?

> BTW, I can apply this PR via patchtester only if I don't apply this one:
> 
> [4.0] PHP 8 Ensure that we have a valid date #36396

That's expected because the other PR includes a change from this PR here, so both PRs touch the same file at the same place and so conflict with each other (which can be resolved later when one of the 2 PRs has been merged).

alikon commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36355">issues.joomla.org/tracker/joomla-cms/36355</a>.</sub>

PhilETaylor commented:
I have removed the Date.php changes in my PR https://github.com/joomla/joomla-cms/pull/36396 and left the Unit Test only as the Date.php issue is fixed in the PR

wilsonge committed:
Cleanup

wilsonge commented on libraries/src/Filesystem/Path.php:
```diff
@@ -216,9 +216,7 @@ public static function clean($path, $ds = DIRECTORY_SEPARATOR)
-		$path = trim($path);
-
-		if (empty($path))
+		if (empty($path) || trim($path) === '')
```
This looks wrong. We've already called is_string above. There's no way this can be null

richard67 replied:
@wilsonge `empty($path)` still catches the empty sting before trimming, too, so it is not absolutely nonsense here. But I think too it's not really necessary. A matter of taste maybe.

frostmakk replied:
I get deprecated message twice when opening Joomla Update.

![image](https://user-images.githubusercontent.com/25645600/148660032-3f5cf209-df7b-43cc-a527-f2e049349aa7.png)

HLeithner replied:
the `if` mentioned by @wilsonge does not evaluate on `null` because `!empty(null)` false in this case.
2 questions here, why does the updater calls this function with `null` and why don't we throw an exception in this case?

The docblock says we only except `string` here

wilsonge replied:
So I guess better fix is to make the first statement an or?

HLeithner replied:
I think `if (!\is_string($path))` for the first if should be enough and `if (trim($path) === '')` for the second if.

But this wouldn't fix the updater.

frostmakk replied:
**What if we left this file as is, and instead changed the default null values to empty strings in
 ..\administrator\components\com_joomlaupdate\src\Model\UpdateModel.php
Could that be an acceptable solution?**

![image](https://user-images.githubusercontent.com/25645600/149297674-6a683476-9402-4ac8-bd2b-20c765fd5bd0.png)

frostmakk replied:
**Or maybe an extra if inside the foreach as an alternative?**

![image](https://user-images.githubusercontent.com/25645600/149319782-339dc98f-dea4-49e5-ab63-3b556989ca3e.png)

wilsonge replied:
The latter I think is better (checking for null) because file_exists on an empty string might return true

richard67 replied:
But then I would not make a nested if but combine the conditions:

```
if ($file !== null && File::exists($file))
{
	File::delete($file);
}
```

frostmakk replied:
Made a separate PR #36684  for libraries/src/Filesystem/Path.php, and stripping this PR down as we go...

wilsonge commented:
I've pulled through some of the less controversial fixes into a direct commit (credit maintained) so we have most of these fixes into 4.0.6. Some of these changes just seem weird in places though (e.g. I'm not sure that JLanguage::hasKey with null should be treated as anything other than a `InvalidArgumentException` 🤷‍♂️ Let's see what's left if we merge in conflicts here and take it from there I think.

zero-24 committed:
Merge branch '4.0-dev' into php81-lib

zero-24 commented:
I have just fixed the merge conflicts so the checking can start.

PhilETaylor commented on libraries/src/Date/Date.php:
```diff
@@ -119,6 +119,12 @@ public function __construct($date = 'now', $tz = null)
 		// Force UTC timezone for correct time handling
 		date_default_timezone_set('UTC');
 
+		// If called with null $date
+		if ($date === null)
+		{
+			$date = 'now';
+		}
+
```
the Date.php changes are tested in PR #36396 by the new audit test, and seem right

PhilETaylor commented on libraries/src/Form/Field/ComponentlayoutField.php:
```diff
@@ -75,7 +75,11 @@ protected function getInput()
 		{
 			$template_style_id = $this->form->getValue('template_style_id', 0);
-			$template_style_id = (int) preg_replace('#\W#', '', $template_style_id);
+
+			if ($template_style_id !== null)
+			{
+				$template_style_id = (int) preg_replace('#\W#', '', $template_style_id);
+			}
```
I would have just gone for 

`$template_style_id = (int) preg_replace('#\W#', '', $template_style_id ?? '');`

PhilETaylor commented on libraries/src/Form/Field/ModulelayoutField.php:
```diff
@@ -76,7 +76,11 @@ protected function getInput()
 		{
 			$template_style_id = $this->form->getValue('template_style_id', 0);
-			$template_style_id = (int) preg_replace('#\W#', '', $template_style_id);
+
+			if ($template_style_id !== null)
+			{
+				$template_style_id = (int) preg_replace('#\W#', '', $template_style_id);
+			}
```
same here - easier to add `??` than another 5 lines of code

PhilETaylor commented on libraries/src/Language/Language.php:
```diff
@@ -1180,6 +1180,11 @@ public function getUsed()
 	 */
 	public function hasKey($string)
 	{
+		if ($string === null)
+		{
+			$string = '';
+		}
+
 		$key = strtoupper($string);
```
here too - just use ?? and now adding another 5 loc.

frostmakk committed:
Review updates

zero-24 committed:
Merge branch '4.0-dev' into php81-lib

frostmakk committed:
Merge branch 'joomla:4.0-dev' into php81-lib

frostmakk committed:
CleanUp some merged parts

frostmakk committed:
More cleanup

Removing files that has stopped producing notices probably due to changes elswhere.

frostmakk committed:
Merge branch 'joomla:4.0-dev' into php81-lib

HLeithner commented:
Sorry to say but all this `?? ''` within a function call is only a band aid it looks like we don't know the input and just ignore the fact that we didn't validated the input correctly.

PhilETaylor commented:
Sorry to say I told you so... try cleaning up docblocks for a day and you will see the most Joomla methods dont know what data types they are actually handling!! Its a complete mess!

HLeithner commented:
Yeah that's my fear... but doesn't mean we should continue todo so right?
We want type hinting for all function signatures for Joomla 5. (we will see if we are able to get them) but having a 4 line block at the beginning that can easily be deleted is better then having hard to find `?? ''` within the code (without comment)

PhilETaylor commented:
unfortunately Joomla is scrambling to meet the demands of 2022 when people are running PHP 8.1 (because Joomla System requirements only specify a LOWER PHP number and not an UPPER PHP number) and so these quick fixes are .. quicker... than a complete code review. 

This kind of work should have been done during the PHP 8.1 RC stages... not, after PHP 8.1.1 stable was released. 

Joomla, as ever, when it comes to pure PHP development is behind the curve. This is the root problem. We need more people focused on code, and less on features and pixels.

HLeithner commented:
true, but doesn't mean that we should make it harder to see the problem in future refactoring.

frostmakk committed:
Merge branch 'joomla:4.0-dev' into php81-lib

wilsonge commented on libraries/src/Date/Date.php:
```diff
@@ -119,6 +119,12 @@ public function __construct($date = 'now', $tz = null)
 		date_default_timezone_set('UTC');
 
+		// If called with null $date
+		if ($date === null)
```
Do we still have any instances of this left? I fixed one in the setLastVisitDate function - I've not seen any others (yet)

frostmakk replied:
I cant find anyone at the moment. 
Reverting this, and if @PhilETaylor knows about some he can put this back into #36396

wilsonge commented on libraries/src/Language/Language.php:
```diff
@@ -1180,7 +1180,7 @@ public function getUsed()
 	public function hasKey($string)
 	{
-		$key = strtoupper($string);
+		$key = strtoupper($string ?? '');
```
This is probably a bad fix - there shouldn't ever be null being passed to this function - if there is - then we should fix it to not be called first

frostmakk replied:
This notice pop up in several places.
Here is a Stack trace from creating a new Random Image module:

PHP Deprecated:  strtoupper(): Passing null to parameter #1 ($string) of type string is deprecated in ..\libraries\src\Language\Language.php on line 1183
PHP Stack trace:
PHP   1. {main}() ..\administrator\index.php:0
PHP   2. require_once() ..\administrator\index.php:32
PHP   3. Joomla\CMS\Application\CMSApplication->execute() ..\administrator\includes\app.php:63
PHP   4. Joomla\CMS\Application\AdministratorApplication->doExecute() ..\libraries\src\Application\CMSApplication.php:278
PHP   5. Joomla\CMS\Application\AdministratorApplication->dispatch($component = *uninitialized*) ..\libraries\src\Application\AdministratorApplication.php:186
PHP   6. Joomla\CMS\Component\ComponentHelper::renderComponent($option = 'com_modules', $params = *uninitialized*) ..\libraries\src\Application\AdministratorApplication.php:143
PHP   7. Joomla\CMS\Dispatcher\ComponentDispatcher->dispatch() ..\libraries\src\Component\ComponentHelper.php:389
PHP   8. Joomla\CMS\MVC\Controller\BaseController->execute($task = 'display') ..\libraries\src\Dispatcher\ComponentDispatcher.php:146
PHP   9. Joomla\Component\Modules\Administrator\Controller\DisplayController->display($cachable = *uninitialized*, $urlparams = *uninitialized*) ..\libraries\src\MVC\Controller\BaseController.php:730
PHP  10. Joomla\CMS\MVC\Controller\BaseController->display($cachable = *uninitialized*, $urlparams = *uninitialized*) ..\administrator\components\com_modules\src\Controller\DisplayController.php:122
PHP  11. Joomla\Component\Modules\Administrator\View\Module\HtmlView->display($tpl = *uninitialized*) ..\libraries\src\MVC\Controller\BaseController.php:692
PHP  12. Joomla\Component\Modules\Administrator\View\Module\HtmlView->addToolbar() ..\administrator\components\com_modules\src\View\Module\HtmlView.php:78
PHP  13. Joomla\CMS\Language\Language->hasKey($string = NULL) ..\administrator\components\com_modules\src\View\Module\HtmlView.php:156
PHP  14. strtoupper($string = NULL) ..\libraries\src\Language\Language.php:1183

bembelimen replied:
Agree, probably "early return" when it's empty?

frostmakk committed:
Revert Path.php

frostmakk committed:
Revert Date.php

Commit:
Merge branch '4.1-dev' into php81-lib

Commit:
Merge branch '4.1-dev' into php81-lib

bembelimen commented on libraries/src/Table/Module.php:
```diff
@@ -157,7 +157,7 @@ public function check()
 
 		// Prevent to save too large content > 65535
-		if ((\strlen($this->content) > 65535) || (\strlen($this->params) > 65535))
+		if ((\strlen($this->content ?? '') > 65535) || (\strlen($this->params) > 65535))
```
As we're checking for "too large" we can add the check in front?

frostmakk commented:
Sorry folks. I don't have time and space in my life for this right now. Please do what you want with this PR, or just close it.
